api/src/test/java/org/openmrs/api/EncounterServiceTest.java

```diff
@@ -98,7 +99,127 @@
+		ObsService os = Context.getObsService();
+
+		Encounter enc = es.getEncounter(100);
+
+		Obs o = os.getObs(101);
+		o.setValueText("Obs value updated");
+
+		es.saveEncounter(enc);
+		Context.evictFromSession(enc);
+		Context.flushSession();
+
+		enc = es.getEncounter(100);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		Obs oParent = os.getObs(100);
+		final Obs editedObs = os.getObs(101);
+		Obs o2 = os.getObs(102);
+		Obs o3 = getNewVersionOfEditedObs(oParent,editedObs);
+
+		assertEquals(1,obsAtTopLevelUpdated.size());
+		assertEquals(3, oParent.getGroupMembers(true).size());
+		assertTrue(editedObs.getVoided());
+		assertFalse(oParent.getVoided());
+		assertFalse(o2.getVoided());
+
+		assertNotNull(o3);
+		assertFalse(o3.getVoided());
+		assertEquals("Obs value updated", o3.getValueText());
+	}
+
+	@Test
+	@Verifies(value = "should update existing encounter correctly when a new obs is added to parent obs", method = "saveEncounter(Encounter)")
+	public void saveEncounter_shouldUpdateExistingEncounterWhenNewObsIsAddedToParentObs() throws Exception {
+		executeDataSet(ENC_OBS_HIERARCHY_DATA_XML);
+		ConceptService cs = Context.getConceptService();
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		Encounter enc = es.getEncounter(100);
+
+		Obs o3 = new Obs();
+		o3.setConcept(cs.getConcept(3));
+		o3.setDateCreated(new Date());
+		o3.setCreator(Context.getAuthenticatedUser());
+		o3.setLocation(new Location(1));
+		o3.setObsDatetime(new Date());
+		o3.setPerson(Context.getPersonService().getPerson(3));
+		o3.setValueText("third obs value text");
+		o3.setEncounter(enc);
+
+		Obs oParent = os.getObs(100);
+		oParent.addGroupMember(o3);
+
+		es.saveEncounter(enc);
+
+		Context.evictFromSession(enc);
+		Context.flushSession();
+
+		enc = es.getEncounter(100);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		assertEquals(1,obsAtTopLevelUpdated.size());
+
+		assertEquals(3, obsAtTopLevelUpdated.iterator().next().getGroupMembers(true).size());
+
+		oParent = os.getObs(100);
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(101)));
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(102)));
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(o3.getObsId())));
+	}
+
+
+	@Test
+	@Verifies(value = "should update existing encounter correctly when an existing obs is removed from parent obs", method = "saveEncounter(Encounter)")
+	public void saveEncounter_shouldSaveEncounterWhenTopLevelObsIsUpdatedByRemovingChildObs() throws Exception {
+		executeDataSet(ENC_OBS_HIERARCHY_DATA_XML);
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		Encounter enc = es.getEncounter(100);
+		Obs oParent = os.getObs(100);
+		Obs oChild = os.getObs(101);
+
+		oParent.removeGroupMember(oChild);
+
+		es.saveEncounter(enc);
+
+		Context.evictFromSession(enc);
```

Comment:
For this particular test when I change to flushSession() and then clearSession() it fails, I personally think these are the correct calls in this order and not evicting the encounter and flushing after. Evicting means remove from the without persisting any changes, is this what you want to happen? I'd think what you want is to push the changes to the DB, clear the whole session so that all entities get reloaded with their new saved state.